server/aesdsocket.c

```diff
@@ -0,0 +1,377 @@
+        }
+
+        /*
+         *  Reference for  inet_ntop : https://man7.org/linux/man-pages/man3/inet_ntop.3.html
+         */ 
+        if(NULL == inet_ntop(AF_INET, get_in_addr((struct sockaddr*)&client_addr),ip_addr, sizeof(ip_addr)))
+        {
+            perror("inet_ntop()");
+            error_handler();
+            exit(ERROR_CODE);
+        }
+        else
+        {
+            syslog(LOG_DEBUG,"Accepted connection from %s", ip_addr);
+        }
+
+        char *data_buf = (char*)calloc(initial_alloc_size,sizeof(char));
+        if(NULL == data_buf)
+        {
+            perror("calloc()");
+            error_handler();
+            exit(ERROR_CODE);
+        }
+
+        /*
+         * For completing any open connection operations i.e., resv() and send(), need to block the signal using sigprocmask()
+         * Blocking doesn't mean we miss the signal, the signal waits until its unblocked and gets serviced by kernel. 
+         */
+		if( -1 == sigprocmask(SIG_BLOCK, &sock_set, NULL))
+        {
+			perror("sigprocmask()");
+			error_handler();
+			exit(ERROR_CODE);
+        }
+
+        while((bytes_read = recv(client_sockfd,data_buf+present_location,initial_alloc_size,0)) > 0)
+        {
+
+            if(-1 == bytes_read)
+            {
+                perror("resv()");
+                error_handler();
+                exit(ERROR_CODE);          
+            }            
+            /*
+             * Reference to strchr function : https://man7.org/linux/man-pages/man3/strchr.3.html
+            */
+            if(strchr(data_buf ,'\n') != NULL)
+            {
+                break;
+            }
+				
+			//If new line is not received, need to keep incrementing the present location to avoid overwritting the previous data received
+			present_location+=bytes_read;
+            extra_alloc++;
+            data_buf = (char*)realloc(data_buf,(extra_alloc*initial_alloc_size)*sizeof(char));
+			if(NULL == data_buf)
+			{
+				syslog(LOG_ERR,"Error: realloc()");
+                free(data_buf);
+                error_handler();
+                exit(ERROR_CODE);
+			}
+        }
+        if(-1 == bytes_read)
+        {
+            perror("resv()");
+            error_handler();
+            exit(ERROR_CODE);          
+        }
+
+        //Write the data to /var/tmp/aesdsocketdata
+        if(-1 == write(data_file_fd,data_buf,strlen(data_buf)))
+        {
+			perror("write()");
+			error_handler();
+			exit(ERROR_CODE);           
+        }
+
+        //Need total data because we need to send back the whole data received till now
+        total_data_len += strlen(data_buf);
+
+        /* To make sure we sent the data from start of the file
+         * Move the cursor to start
+		 */
+        lseek(data_file_fd, 0, SEEK_SET);
+
+        //create a buffer to send data with total size available at /var/tmp/aesdsocketdata
+        char *send_to_client_buf = (char*)calloc(total_data_len,sizeof(char));
```

Comment:
@Rishikesh38 what happens when enough packets have been sent that `total_data_len` is GB in size?  Will this code meet assignment requirements?

Reply:
I had to handle the total size of all packets sent will be less than the size of the available RAM for the process heap